src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
+        } else if (index == 0) {
+            Node<T> insertNode = new Node<>(null,value,head);
+            head.setPrev(insertNode);
+            head = insertNode;
+        } else {
+            Node<T> refNode = findNodeByIndex(index);
+            Node<T> insertNode = new Node<>(refNode.getPrev(),value,refNode);
+            refNode.getPrev().setNext(insertNode);
+            refNode.setPrev(insertNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findNodeByIndex(index).getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> oldNode = findNodeByIndex(index);
+        T oldValue = oldNode.getValue();
+        oldNode.setValue(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index is out of range");
+        }
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> removeNode = head; removeNode != null; removeNode = removeNode.getNext()) {
+                if (removeNode.getValue() == null) {
+                    unlink(removeNode);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> removeNode = head; removeNode != null; removeNode = removeNode.getNext()) {
+                if (object.equals(removeNode.getValue())) {
+                    unlink(removeNode);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> unlinkNode) {
+        final T element = unlinkNode.getValue();
+        final Node<T> next = unlinkNode.getNext();
+        final Node<T> prev = unlinkNode.getPrev();
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.setNext(next);
+            unlinkNode.setPrev(null);
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.setPrev(prev);
+            unlinkNode.setNext(null);
+        }
+        unlinkNode.setValue(null);
+        size--;
+        return element;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index is out of range");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
```

Comment:
[checklist](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html): Don’t iterate your nodes from the head to tail or vice versa.